File: stockpro-frontend/src/services/warehouseService.js
import { warehouseApi } from './api';

export const warehouseService = {
    // ==================== Warehouse CRUD ====================

    // Get all active warehouses
    getAllWarehouses: async () => {
        const response = await warehouseApi.get('/warehouse');
        return response.data;
    },

    // Get a warehouse by ID
    getWarehouseById: async (id) => {
        const response = await warehouseApi.get(`/warehouse/${id}`);
        return response.data;
    },

    // Create a new warehouse
    createWarehouse: async (warehouseData) => {
        const response = await warehouseApi.post('/warehouse', warehouseData);
        return response.data;
    },

    // Update a warehouse
    updateWarehouse: async (id, warehouseData) => {
        const response = await warehouseApi.put(`/warehouse/${id}`, warehouseData);
        return response.data;
    },

    // Delete (soft) a warehouse
    deleteWarehouse: async (id) => {
        const response = await warehouseApi.delete(`/warehouse/${id}`);
        return response.data;
    },

    // ==================== Stock Level Management ====================

    // Get stock level for a specific product in a warehouse
    getStockLevel: async (warehouseId, productId) => {
        const response = await warehouseApi.get(`/warehouse/${warehouseId}/stock/${productId}`);
        return response.data;
    },

    // Get all stock levels in a warehouse
    getStockByWarehouse: async (warehouseId) => {
        const response = await warehouseApi.get(`/warehouse/${warehouseId}/stock`);
        return response.data;
    },

    // Get stock levels for a product across all warehouses
    getStockByProduct: async (productId) => {
        const response = await warehouseApi.get(`/warehouse/stock/product/${productId}`);
        return response.data;
    },

    // Add stock to a warehouse
    addStock: async (stockData) => {
        const response = await warehouseApi.post('/warehouse/stock/add', stockData);
        return response.data;
    },

    // Deduct stock from a warehouse
    deductStock: async (stockData) => {
        const response = await warehouseApi.post('/warehouse/stock/deduct', stockData);
        return response.data;
    },

    // Reserve stock (soft-hold for pending orders)
    reserveStock: async (stockData) => {
        const response = await warehouseApi.post('/warehouse/stock/reserve', stockData);
        return response.data;
    },

    // Release a previous stock reservation
    releaseReservation: async (stockData) => {
        const response = await warehouseApi.post('/warehouse/stock/release', stockData);
        return response.data;
    },

    // Transfer stock between warehouses
    transferStock: async (transferData) => {
        const response = await warehouseApi.post('/warehouse/stock/transfer', transferData);
        return response.data;
    },

    // Get all low stock items
    getLowStockItems: async (threshold = 10) => {
        const response = await warehouseApi.get(`/warehouse/stock/low-stock?threshold=${threshold}`);
        return response.data;
    }
};

File: StockPro-Demo-Backend/microservices/warehouse-service/src/main/java/com/stockpro/warehouse/service/WarehouseService.java
package com.stockpro.warehouse.service;

import com.stockpro.warehouse.dto.StockLevelResponseDTO;
import com.stockpro.warehouse.dto.StockTransferRequestDTO;
import com.stockpro.warehouse.dto.StockUpdateRequestDTO;
import com.stockpro.warehouse.dto.WarehouseRequestDTO;
import com.stockpro.warehouse.dto.WarehouseResponseDTO;

import java.util.List;

public interface WarehouseService {

    // --- Warehouse CRUD ---
    WarehouseResponseDTO createWarehouse(WarehouseRequestDTO requestDTO);
    WarehouseResponseDTO getWarehouseById(Integer id);
    List<WarehouseResponseDTO> getAllActiveWarehouses();
    WarehouseResponseDTO updateWarehouse(Integer id, WarehouseRequestDTO requestDTO);
    void deleteWarehouse(Integer id); // Soft Delete

    // --- Stock Level Management ---
    StockLevelResponseDTO getStockLevel(Long warehouseId, Long productId);
    List<StockLevelResponseDTO> getStockByWarehouse(Long warehouseId);
    List<StockLevelResponseDTO> getStockByProduct(Long productId);
    StockLevelResponseDTO addStock(StockUpdateRequestDTO request);
    StockLevelResponseDTO deductStock(StockUpdateRequestDTO request);
    StockLevelResponseDTO reserveStock(StockUpdateRequestDTO request);
    StockLevelResponseDTO releaseReservation(StockUpdateRequestDTO request);
    void transferStock(StockTransferRequestDTO request);
    List<StockLevelResponseDTO> getLowStockItems(int threshold);
}